File: tl/src/main/java/com/github/badoualy/telegram/tl/api/TLInputGeoPoint.java
package com.github.badoualy.telegram.tl.api;

import com.github.badoualy.telegram.tl.TLContext;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import static com.github.badoualy.telegram.tl.StreamUtils.readDouble;
import static com.github.badoualy.telegram.tl.StreamUtils.writeDouble;

/**
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
public class TLInputGeoPoint extends TLAbsInputGeoPoint {
    public static final int CLASS_ID = 0xf3b7acc9;

    protected double lat;

    protected double _long;

    public TLInputGeoPoint() {
    }

    public TLInputGeoPoint(double lat, double _long) {
        this.lat = lat;
        this._long = _long;
    }

    @Override
    public void serializeBody(OutputStream stream) throws IOException {
        writeDouble(lat, stream);
        writeDouble(_long, stream);
    }

    @Override
    @SuppressWarnings("unchecked")
    public void deserializeBody(InputStream stream, TLContext context) throws IOException {
        lat = readDouble(stream);
        _long = readDouble(stream);
    }

    @Override
    public String toString() {
        return "inputGeoPoint#f3b7acc9";
    }

    @Override
    public int getClassId() {
        return CLASS_ID;
    }

    public double getLat() {
        return lat;
    }

    public void setLat(double lat) {
        this.lat = lat;
    }

    public double getLong() {
        return _long;
    }

    public void setLong(double _long) {
        this._long = _long;
    }

    @Override
    public final boolean isEmpty() {
        return false;
    }

    @Override
    public final boolean isNotEmpty() {
        return true;
    }

    @Override
    public final TLInputGeoPoint getAsInputGeoPoint() {
        return this;
    }
}

File: tl/src/main/kotlin/com/github/badoualy/telegram/tl/api/TLInputGeoPoint.kt
package com.github.badoualy.telegram.tl.api

import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_DOUBLE
import com.github.badoualy.telegram.tl.serialization.TLDeserializer
import com.github.badoualy.telegram.tl.serialization.TLSerializer
import java.io.IOException

/**
 * inputGeoPoint#f3b7acc9
 *
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
class TLInputGeoPoint() : TLAbsInputGeoPoint() {
    var lat: Double = 0.0

    var _long: Double = 0.0

    private val _constructor: String = "inputGeoPoint#f3b7acc9"

    override val constructorId: Int = CONSTRUCTOR_ID

    constructor(lat: Double, _long: Double) : this() {
        this.lat = lat
        this._long = _long
    }

    @Throws(IOException::class)
    override fun serializeBody(tlSerializer: TLSerializer) = with (tlSerializer)  {
        writeDouble(lat)
        writeDouble(_long)
    }

    @Throws(IOException::class)
    override fun deserializeBody(tlDeserializer: TLDeserializer) = with (tlDeserializer)  {
        lat = readDouble()
        _long = readDouble()
    }

    override fun computeSerializedSize(): Int {
        var size = SIZE_CONSTRUCTOR_ID
        size += SIZE_DOUBLE
        size += SIZE_DOUBLE
        return size
    }

    override fun toString() = _constructor

    override fun equals(other: Any?): Boolean {
        if (other !is TLInputGeoPoint) return false
        if (other === this) return true

        return lat == other.lat
                && _long == other._long
    }
    companion object  {
        const val CONSTRUCTOR_ID: Int = 0xf3b7acc9.toInt()
    }
}
